### DISCUSSION TITLE
VexRiscV-SMP/FPU Support

### DISCUSSION THREAD

**Discussion by pottendo** [zeldin/OrangeCart#5] [RVCop64]
Hi,
managed to build the RVCop64 bitstream to feature the VexRiscV-SMP CPU. One can enable FPU (single core) or two cores. Both at the same time won't fit into the FPGA.
Now even Linux (from here: https://github.com/litex-hub/linux-on-litex-vexriscv) boots - see attached bootlog.
Of course this has to be tweaked to get the root-fs on the SDCard and maybe some more optimization to save RAM to be useful; but that's for another session for another day...

Biggest hurdle was to get the build-system running; some minor changes in 'bitstream.py' and 'basesoc.py' and that's it.
I needed to map the C64 to a different memory location, as this CPU expects the boot-vector at 0x00000000. With that I confirmed that memory access from the cart to the c64 mem still works; this is also something to be checked on the Linux side... 
As far as I could see, also FPU instructions work; more testing to be done, though.

The downside: Zephyr needs more tweaking. I wasn't able to get this going quickly; I don't exactly understand the device-tree concept yet, so any help here would be highly appreciated.

Any comments are welcome,
pottendo

[Linux-FPU-Boot.log](https://github.com/zeldin/OrangeCart/files/10146846/Linux-FPU-Boot.log)

**Comment by pottendo** [comment]
Hi,
some progress to report:
Linux boots into the console - see attached bootlog.
Here you see a dual-core CPU:
````
Welcome to Buildroot
buildroot login: root
                   __   _
                  / /  (_)__  __ ____ __
                 / /__/ / _ \/ // /\ \ /
                /____/_/_//_/\_,_//_\_\
                      / _ \/ _ \
   __   _ __      _  _\___/_//_/         ___  _
  / /  (_) /____ | |/_/__| | / /____ __ / _ \(_)__ _____  __
 / /__/ / __/ -_)>  </___/ |/ / -_) \ // , _/ (_-</ __/ |/ /
/____/_/\__/\__/_/|_|____|___/\__/_\_\/_/|_/_/___/\__/|___/
                  / __/  |/  / _ \
                 _\ \/ /|_/ / ___/
                /___/_/  /_/_/
  32-bit RISC-V Linux running on LiteX / VexRiscv-SMP with USB host support.

login[86]: root login on 'console'
root@buildroot:~# cat /proc/cpuinfo 
processor	: 0
hart		: 0
isa		: rv32ima
mmu		: sv32
mvendorid	: 0x0
marchid		: 0x0
mimpid		: 0x0

processor	: 1
hart		: 1
isa		: rv32ima
mmu		: sv32
mvendorid	: 0x0
marchid		: 0x0
mimpid		: 0x0

root@buildroot:~#
````

There's some .config tweaking necessary in order to get the speed and responsiveness somewhat acceptable. But as a PoC it works.

Also the memory seems a bit odd:
[    0.000000] Memory: 7156K/16384K available (6566K kernel code, 386K rwdata, 1110K rodata, 228K init, 198K bss, 9228K reserved, 0K cma-reserved)

Not sure why there's 9MB reserved, or what this actually means - it seems that not the full 16MB are used:
```
root@buildroot:~# free -h
              total        used        free      shared  buff/cache   available
Mem:           7.2M        4.3M      436.0K       16.0K        2.4M        1.7M
root@buildroot:~# cat /proc/meminfo 
MemTotal:           7380 kB
MemFree:             544 kB
MemAvailable:       1752 kB
[...]
```

Now I need to find out how to access the C64... 
Feedback is welcome!

bye, pottendo 

PS: I got Zephyr's console working on a single-core/FPU bitstream - but all time-related functions don't work. So I struggle to get the HW timer connected to the operating system. Any ideas are welcome!


[Linux-DualCore-Boot.log](https://github.com/zeldin/OrangeCart/files/10195255/Linux-DualCore-Boot.log)

**Comment by pottendo** [comment]
hi,

I've optimized memory a bit (@Linux kernel level) to get some room for applications and played a bit with FPU. It's been a hassle to get the right compiler done, but with buildroot this is fairly straightforward, once one is gets used to it a bit...
Booting from the sdcard is a bit slow-ish, same to load large applications (e.g. C++ std share lib)... so I fear it won't be more than a PoC.

FPU makes a difference:
without: https://photos.app.goo.gl/Vp8G3WNZk6D5hxou8
Double Precision FPU: https://photos.app.goo.gl/ygEqyVG7yZ9ZQ7VU9

Unfortunately DualCore & FPU won't fit into our FPGA, but at least one can use FPU (excl.-)or DualCore (only with Linux yet).

Still need to find out to boot Zephyr; this would be a great lightweight alternative.

happy hacking, pottendo

PS: will provide my changes as soon as I've cleaned up and support build for both CPUs to be selected from commandline.

**Comment by pottendo** [comment]
Hi, here a tiny xmas update:
I got Zephyr booting incl. timer and (double precision) FPU support on a Single Core / FPU VexRiscV-smp CPU; this really rocks now!
The changes are not many as Zephyr has a lot on-board already. I've even added 2 board variants, so one can select on during application build for which bitstream to build; probably I'll add another one for DualCore/without FPU.
I've confirmed access to some of the I/Os (leds & SPI/SDCard read-only) and C64 address range. With the adapted phase_shift all access is clean & fine on my breadbox.
Once I've cleaned up, learned 'git' better I'll push my changes upstream.
Merry Christmas!
pottendo

**Comment by pottendo** [comment]
Dear fellow hackers,
find the updstream commits here: https://github.com/pottendo/RVCop64-pottendo
including pre-built bitstreams.
have fun, pottendo

**Reply by LeifBloomquist** [reply]
Very interesting!   So I was able to load this bitstream just fine with dfu-util -a 0 -R -D RVCop64-dc-v1.0.bit

I get the modified BASIC startup screen, but when I enter RVTERM the screen background goes black, but there is no data from the FPGA or anything when I type.  Is there another step I need to take?   (Sorry for the beginner question!)

**Comment by pottendo** [comment]
Hi, 
my Zephyr fork is now online: https://github.com/pottendo/zephyr-pottendo
FPU & SMP(Dual Core) is only usable in conjunction with my bitstreams (https://github.com/pottendo/RVCop64-pottendo). Using the board (option `-b orangecart` when building with `west`) works with the default bitstream of RVCop64.
Zephyr@DualCore is probably not yet using the both cores... needs further devicetree changes (I assume).
I'll add some concrete example how to use it later.
Any feedback is welcome,
pottendo

**Comment by pottendo** [comment]
Hi,
Sorry, yes, I need to add that the console is exclusively configured for USB to the host PC.
Probably I should provide the prebuilt bitstreams with the C64 console. 

So, connect the USB cable to start the cartridge (c64 should be off as well) use something like 
  `litex_term /dev/ttyACM0` 
on your Linux host, and it should boot into the Litex BIOS.
To load programs use the option `--kernel=bla.bin`
Hope this helps,
pottendo

**Comment by pottendo** [comment]
hi,
some update: my [Zephyr tree](https://github.com/pottendo/zephyr-pottendo) is now updated to the new shiny Zephyr 3.3.0.
Usable with bitstreams configured for CPUs
- **vexriscv-smp, single core / FPU**. 
This needs the most recent vexriscv-smp (for FPU fixes), so make sure you update your RVCop64 to get the latest before building your own bitstream.
See [here](https://photos.app.goo.gl/MVwJAkooZyhqnms49) for mandelbrot-set zoomer output on the C64 (sorry for the poor video quality).
- **vexriscv-smp, dual core**. 
Confirmed to use both cores (no FPU) on Zephyr also now.

To build the bitstreams, you need to use my RVCop64 fork from [here](https://github.com/pottendo/RVCop64-pottendo).
Now I'm working on ppp via the second uart to connect to a wifi-network - via an ESP32; Still need to find a proper modem/ppp gateway implementation for the esp32. [This](https://github.com/ep00ch/esp32-at-ppp/) looks promising, but a bit 'oldish' unfortunately, so building still fails for me... 

Both bitstream can also boot Linux - including access to C64 mem using mmap; see [here](https://github.com/pottendo/zephyr-pottendo/blob/main/samples/mandelbrot/src/c64-lib.cpp) how this can be done.

happy hacking, pottendo
